File: lib/py/etl/paradicms_etl/models/property.py
from abc import abstractmethod

from rdflib import URIRef

from paradicms_etl.model import Model
from paradicms_etl.models.images_mixin import ImagesMixin


class Property(Model, ImagesMixin):
    @property
    @abstractmethod
    def filterable(self) -> bool | None:
        raise NotImplementedError

    @property
    @abstractmethod
    def label(self) -> str:
        raise NotImplementedError

    @property
    @abstractmethod
    def order(self) -> int | None:
        raise NotImplementedError

    @property
    @abstractmethod
    def range(self) -> URIRef | None:  # noqa: A003
        raise NotImplementedError

    @property
    @abstractmethod
    def searchable(self) -> bool | None:
        raise NotImplementedError

File: lib/ts/models/src/Property.ts
import {NamedNode} from "@rdfjs/types";
import {ImagesMixin} from "./ImagesMixin";
import {Model} from "./Model";
import {PropertyGroup} from "./PropertyGroup";
import {PropertyValue} from "./PropertyValue";
import {Text} from "./Text";
import {ThumbnailMixin} from "./ThumbnailMixin";

export interface Property extends Model, ImagesMixin, ThumbnailMixin {
  readonly comment: Text | null;
  readonly filterable: boolean;
  readonly hidden: boolean;
  readonly label: string;
  readonly groups: readonly PropertyGroup[];
  readonly order: number;
  readonly range: NamedNode | null;
  readonly rangeValues: readonly PropertyValue[];
  readonly searchable: boolean;
}
